python/src/iceberg/partition_spec.py

```diff
@@ -0,0 +1,289 @@
+        partition_spec_str = "["
+        for field in self.fields:
+            partition_spec_str += "\n"
+            partition_spec_str += " " + field
+        if len(self.fields) > 0:
+            partition_spec_str += "\n"
+        partition_spec_str += "]"
+        return partition_spec_str
+
+    def compatible_with(self, other):
+        if self.__eq__(other):
+            return True
+
+        if len(self.fields) != len(other.fields):
+            return False
+
+        index = 0
+        for field in self.fields:
+            other_field: PartitionField = other.fields[index]
+            if (
+                field.source_id != other_field.source_id
+                or field.name != other_field.name
+            ):
+                index += 1
+                # TODO: Add transform check
+                return False
+        return True
+
+    def partition_type(self):
+        struct_fields = []
+        # TODO: Needs transform
+        pass
+
+    def escape(self, input_str):
+        try:
+            return parse.urlencode(input_str, encoding="utf-8")
+        except TypeError as e:
+            raise e
+
+    def partition_to_path(self):
+        # TODO: Needs transform
+        pass
+
+    def _generate_unpartitioned_spec(self):
+        return PartitionSpec(
+            schema=Schema(),
+            spec_id=0,
+            part_fields=[],
+            last_assigned_field_id=PARTITION_DATA_ID_START - 1,
+        )
+
+    def unpartitioned(self) -> PartitionSpec:
+        return self._generate_unpartitioned_spec()
+
+    def check_compatibility(self, spec: PartitionSpec, schema: Schema):
+        for field in spec.fields:
+            source_type = schema.find_type(field.source_id)
+            ValidationException().check(
+                source_type != None,
+                f"Cannot find source column for partition field: {field}",
+            )
+            ValidationException().check(
+                source_type.is_primitive,
+                f"Cannot partition by non-primitive source field: {source_type}",
+            )
+            # TODO: Add transform check
+
+    def has_sequential_ids(self, spec: PartitionSpec):
+        index = 0
+        for field in spec.fields:
+            if field.field_id != PARTITION_DATA_ID_START + index:
+                return False
+            index += 1
+        return True
+
+
+class AtomicInteger:
+    # TODO: Move to utils
+    def __init__(self, value=0):
+        self._value = int(value)
+        self._lock = threading.Lock()
+
+    def inc(self, d=1):
+        with self._lock:
+            self._value += int(d)
+            return self._value
+
+    def dec(self, d=1):
+        return self.inc(-d)
+
+    @property
+    def value(self):
+        with self._lock:
+            return self._value
+
+    @value.setter
+    def value(self, v):
+        with self._lock:
+            self._value = int(v)
+            return self._value
+
+
+class Builder(object):
```

Comment:
But we cannot enforce immutability of the original object in python and it is always mutable. So it misses the main benefit of builder pattern in Java. So I am not sure if there is a confusion for python users that the original object is expected to be immutable or not.

A builder class here seems to be a boilerplate class with an extra layer to prevent calling the target object's methods but it won't prevent the field in the target object being mutated.

There are also multiple other places using builder pattern in iceberg. It is better that we keep this consistent (either use or not use builder class). I will start a thread in the slack channel to get more options about it from python community about builder class.

Reply:
Discussion continued here on Slack: https://apache-iceberg.slack.com/archives/C029EE6HQ5D/p1635538598009300